components/director/internal/tenantfetcher/service.go

```diff
@@ -88,82 +131,135 @@ func (s Service) SyncTenants() error {
 	}
 
 	tenantsToCreate, err := s.getTenantsToCreate(lastConsumedTenantTimestamp)
-
 	if err != nil {
 		return err
 	}
-	tenantsToCreate = s.dedupeTenants(tenantsToCreate)
 
 	tenantsToDelete, err := s.getTenantsToDelete(lastConsumedTenantTimestamp)
 	if err != nil {
 		return err
 	}
 
-	totalNewEvents := len(tenantsToCreate) + len(tenantsToDelete)
-	log.C(ctx).Printf("Amount of new events: %d", totalNewEvents)
-	if totalNewEvents == 0 {
-		return nil
+	runtimesToMove, err := s.getRuntimesToMoveByLabel(lastConsumedTenantTimestamp)
+	if err != nil {
+		return err
 	}
 
-	deleteTenantsMap := make(map[string]model.BusinessTenantMappingInput)
-	for _, ct := range tenantsToDelete {
-		deleteTenantsMap[ct.ExternalTenant] = ct
-	}
+	tenantsToCreate = s.dedupeTenants(tenantsToCreate)
+	tenantsToCreate = s.excludeTenants(tenantsToCreate, tenantsToDelete)
 
-	for i := len(tenantsToCreate) - 1; i >= 0; i-- {
-		if _, found := deleteTenantsMap[tenantsToCreate[i].ExternalTenant]; found {
-			tenantsToCreate = append(tenantsToCreate[:i], tenantsToCreate[i+1:]...)
-		}
+	totalNewEvents := len(tenantsToCreate) + len(tenantsToDelete) + len(runtimesToMove)
+	log.C(ctx).Printf("Amount of new events: %d", totalNewEvents)
+	if totalNewEvents == 0 {
+		return nil
 	}
 
 	tx, err := s.transact.Begin()
 	if err != nil {
 		return err
 	}
 	defer s.transact.RollbackUnlessCommitted(ctx, tx)
-
 	ctx = persistence.SaveToContext(ctx, tx)
 
-	currentTenants, err := s.tenantStorageService.List(ctx)
-	if err != nil {
-		return errors.Wrap(err, "while listing tenants")
-	}
+	actions := make([]*syncTenantsAction, 0)
 
-	currentTenantsMap := make(map[string]bool)
-	for _, ct := range currentTenants {
-		currentTenantsMap[ct.ExternalTenant] = true
+	if len(tenantsToCreate) == 0 && len(tenantsToDelete) == 0 {
```

Comment:
Yeah, they handle it, but it is not visible enough. You have to look a lot deeper in the call hierarchy in order to understand this. IMO, ensuring that you pass only valid/meaningful data on top level method seems to be more appropriate way to go. Also, ... 
- this prevent passing nil values to the action methods. 
- one db call can be avoided _(getting the existing tenants)_

It is also possible to place an explicit check (the same check) at the beginning of each action method. This will make entrypoint method `SyncTenants` more compact, but it will lead to:
 - add extra check for empty slice at the beginning of each action method
 - bring back the previously introduced `currentTenantsSupplier` struct (get tenants if not did it yet) in order to ensure one db call (or none -  for missing create/delete events) is executed to fetch the current tenants. (`createTenants / deleteTenants` need it)

Both approaches have their advantages. If you insist on cleaner `SyncTenants` method we can go with the second approach. ;)